kahuna/public/js/services/label.js

```diff
@@ -2,18 +2,43 @@ import angular from 'angular';
 
 var labelService = angular.module('kahuna.services.label', []);
 
-labelService.factory('labelService', ['$rootScope', '$q', function ($rootScope, $q) {
+labelService.factory('labelService',
+                     ['$rootScope', '$q', 'apiPoll', 'imageAccessor',
+                      function ($rootScope, $q, apiPoll, imageAccessor) {
+
+    function readLabelName(label) {
+        return label.data;
+    }
+
+    function readLabelsName(labels) {
+        return labels.map(readLabelName);
+    }
+
+    function labelsEquals(labelsA, labelsB) {
+        return angular.equals(readLabelsName(labelsA), readLabelsName(labelsB));
+    }
+
+    function untilLabelsEqual(image, expectedLabels) {
+        return image.get().then(apiImage => {
+            const apiLabels = imageAccessor.readLabels(apiImage);
+            if (labelsEquals(apiLabels, expectedLabels)) {
+                return apiImage;
+            } else {
+                return $q.reject();
+            }
+        });
+    }
+
     function remove (image, label) {
-        var existingLabels = image.data.userMetadata.data.labels.data;
+        var existingLabels = imageAccessor.readLabels(image);
         var labelIndex = existingLabels.findIndex(lbl => lbl.data === label);
-
         if (labelIndex !== -1) {
             return existingLabels[labelIndex]
                 .delete()
-                .then(newLabels => {
-                    image.data.userMetadata.data.labels = newLabels;
-                    $rootScope.$emit('image-updated', image, image);
-                    return image;
+                .then(newLabels => apiPoll(() => untilLabelsEqual(image, newLabels.data)))
```

Comment:
> without doing a deep comparison which would be very inefficient

ATM, each component is watching their part (?) of the image resource only, rather than the whole image e.g. https://github.com/guardian/grid/blob/master/kahuna/public/js/edits/labeller.js#L19.

> I'd like to be able to reflect the change on the client-side only

Don't quite understand what you mean...